### PR TITLE
DQN project, first draft

### PR DESCRIPTION
**Pull request by lucasosouza**
Implementation of DQN in Vernon. Can be used as substrate for other embodied AI implementations. 

First draft, not fully functional (runs, but doesn't learn as well). Requires some debugging. 

I will likely merge the first draft, so others can build/work on it, but comments and feedbacks are appreciated and will be incorporated in a second PR. We can also continue the discussion on github even if the PR is merged. Thanks.

### CODE REVIEW

File: projects/dqn/src/data_utils.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                self.last_observation,
+                action,
+                reward,
+                done,
+                next_observation,
+                info
+            )
+            self.last_observation = next_observation
+            yield experience
+        self.reset()
```
Comment (benja-matic) [inline projects/dqn/src/data_utils.py:115]: Move this to line 101? Then eliminate the call to reset() on line 91?

File: projects/dqn/src/algorithms.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        self.device = device
+
+        # Initialize networks
+        input_channels = observation_size[2]
+        self.q_function = q_function_network_class(
+            input_channels=input_channels,
+            num_actions=num_actions
+        )
+        self.q_function.to(self.device)
+        self.target_q_function = q_function_network_class(
```
Comment (benja-matic) [inline projects/dqn/src/algorithms.py:67]: Supposed to be initialized from exact same weights.

File: projects/dqn/src/algorithms.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        self.clip_error = clip_error
+        self.discount_rate = discount_rate
+        self.target_function_update_rate = target_function_update_rate
+        self.target_function_update_frequency = target_function_update_frequency
+
+    def gradient_step(self, experiences):
+        self.q_function_optimizer.zero_grad()
+        output, loss = self.compute_error(experiences)
+        output.backward(loss)
+        self.q_function_optimizer.zero_grad()
```
Comment (benja-matic) [inline projects/dqn/src/algorithms.py:95]: Step, not zero grad

File: projects/dqn/src/algorithms.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        # next_max_q and next_q_values is 128
+        next_max_q = self.target_q_function(next_obs_batch).detach().max(1)[0]
+        next_q_values = not_done_mask * next_max_q
+
+        # Compute the target of the current Q values
+        # rew_batch target_q_values 128
+        target_q_values = rew_batch + (self.discount_rate * next_q_values)
+
+        # Compute Bellman error
+        bellman_error = target_q_values - current_q_values
```
Comment (benja-matic) [inline projects/dqn/src/algorithms.py:156]: TD error

File: projects/dqn/src/algorithms.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        if self.clip_error:
+            bellman_error = bellman_error.clamp(-1, 1)
+
+        # Note: bellman_delta * -1 will be right gradient for minimization
+        loss = bellman_error * -1.0
+
+        return current_q_values, loss
+
+    def post_step(self, num_steps):
+        self.update_target_function(num_steps)
```
Comment (benja-matic) [inline projects/dqn/src/algorithms.py:166]: Call scheduler step

File: projects/dqn/experiments/base.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+"""
+Base GSC Experiment configuration.
+"""
+
+import os
+
+import gym
+
+from nupic.research.frameworks.ray.trainables import DebugTrainable
+from src import DQN, EnvDataset, OffPolicyReinforcementLearningExperiment
```
Comment (lscheinkman) [inline projects/dqn/experiments/base.py:30]: You should not import from `src`, it is confusing. This folder should not be used as the package name. If you are not creating an installable package then i would just remove the `src` folder to avoid confusion. 
See https://packaging.python.org/en/latest/tutorials/packaging-projects/

File: projects/dqn/src/misc.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+
+class Scheduler():
```
Comment (lscheinkman) [inline projects/dqn/src/misc.py:23]: It looks like this file should be called `schedulers.py`

File: projects/dqn/src/rl_experiment.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+        self.logger.debug("---------- End of run epoch ------------")
+        self.logger.debug("")
+
+        self.current_epoch += 1
+        return ret
+
+    def train_epoch(self):
+        # Train loader must be recreated at each step due to different sample subset
+        train_loader = self.create_train_loader()
+        epoch_loss = 0
```
Comment (lscheinkman) [inline projects/dqn/src/rl_experiment.py:164]: If you use a GPU tensor for `epoch_loss` then you can keep the loss returned by `gradient_step` also on GPU and avoid the GPU->CPU transfer in the loop
```
 epoch_loss = torch.tensor(0, device=self.device)
```

File: projects/dqn/src/rl_experiment.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+
+        # Train
+        t1 = time.time()
+        ret["average_loss"] = self.train_epoch()
+        t2 = time.time()
+
+        self.post_epoch()
+
+        self.logger.debug("data collection time: %s", t1 - t0)
+        self.logger.debug("training time: %s", t2 - t1)
```
Comment (lscheinkman) [inline projects/dqn/src/rl_experiment.py:154]: This timing would only work for CPU benchmarking. CUDA is asynchronous.
To measure GPU time you need to use `torch.cuda.Event` and `torch.cuda.synchronize`

**Review by lscheinkman — CHANGES_REQUESTED** [review]
Not sure I understand the urgency to merge this "draft" PR now and create a second PR later.
Why not wait to merge the until  PR is complete?

File: projects/dqn/src/networks.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        """
+        Initialize a deep Q-learning network
+        Arguments:
+            input_channels: number of channel of input.
+                            In DQN, can correspond to number of stacked frames
+            num_actions: number of action-value to output, one-to-one correspondence to
+                         action in game.
+        """
+        super().__init__()
+        self.features = nn.Sequential(
```
Comment (vkakerbeck) [inline projects/dqn/src/networks.py:36]: No activation function? I think in the original paper they use ReLUs

File: projects/dqn/src/algorithms.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+    def load_state_dict(self, state_dict):
+        self.q_function.load_state_dict(state_dict["q_function"])
+        self.target_q_function.load_state_dict(state_dict["target_q_function"])
+        self.q_function_optimizer.load_state_dict(state_dict["q_function_optimizer"])
+        self.exploration_rate_scheduler.load_state_dict(
+            state_dict["exploration_rate_scheduler"])
+
+    def get_next_action(self, last_observation):
+        if np.random.random() < self.exploration_rate_scheduler():
```
Comment (vkakerbeck) [inline projects/dqn/src/algorithms.py:114]: Could also check here whether we are training or evaluating. We only need to be epsilon greedy during training and can act greedy for evaluation.

File: projects/dqn/src/algorithms.py (hunk 2 above)
Comment (lucasosouza) [reply projects/dqn/src/algorithms.py:67]: changed to deepcopy

File: projects/dqn/run.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+from nupic.research.frameworks.vernon.parser_utils import (
+    get_default_parsers,
+    process_args,
+)
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        parents=get_default_parsers(),
+    )
+    parser.add_argument("-e", "--experiment", dest="name", default="default_base",
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/dqn/run.py:36]: I meant `base`, fixed

File: projects/dqn/src/rl_experiment.py (hunk 9 above)
Comment (lucasosouza) [reply projects/dqn/src/rl_experiment.py:154]: Even though it is asynchronous, I still have to wait until all is done to return the loss in `train_epoch`. Even if inside  `collect_data` and `train_epoch` there is parallelization, that should not change the time collected in the outer loop. Not sure I understand your point here.

File: projects/dqn/src/networks.py (hunk 10 above)
Comment (lucasosouza) [reply projects/dqn/src/networks.py:36]: good catch

**Review by subutai — APPROVED** [review]

File: projects/dqn/src/algorithms.py (hunk 11 above)
Comment (lucasosouza) [reply projects/dqn/src/algorithms.py:114]: Good idea. This first implementation only has training though, and the reward accrued in the training is episoded is tracked as a performance metric. I left a reminder in docstring that this should only be used in training and exploration should be removed in evaluation

File: projects/dqn/src/rl_experiment.py (hunk 8 above)
Comment (lucasosouza) [reply projects/dqn/src/rl_experiment.py:164]: Good idea, but only moving an integer shouldn't be that costly. In this case would rather have my gradient step in the algorithm to return an int as the loss and not a tensor so it doesn't get messy when logging or debugging.

File: projects/dqn/experiments/base.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def gym_env(**env_args):
+    return gym.make(**env_args)
+
+
+base = dict()
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/dqn/experiments/base.py:37]: That was to be the default, as seen above. It is just to keep it consistent. All defaults are defined in the dataclasses instead.

File: projects/dqn/experiments/base.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+"""
+Base GSC Experiment configuration.
+"""
+
+import os
+
+import gym
+
+from nupic.research.frameworks.ray.trainables import DebugTrainable
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/dqn/experiments/base.py:29]: Not really

File: projects/dqn/src/data_utils.py (hunk 1 above)
Comment (lucasosouza) [reply projects/dqn/src/data_utils.py:115]: > Not sure I understand the urgency to merge this "draft" PR now and create a second PR later. Why not wait to merge the until PR is complete?

There is no urgency. Is just this was only meant as a reference for Monty, and a skeleton for future RL projects. There is no interest in spending more time debugging and getting it perfect at this point in time, we may revisit later.

File: projects/dqn/experiments/base.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+#  http://numenta.org/licenses/
+#
+
+"""
+Base GSC Experiment configuration.
+"""
+
+import os
+
+import gym
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/dqn/experiments/base.py:27]: added

File: projects/dqn/experiments/base.py [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+dqn_base = dict(
+    logdir=os.path.expanduser("~/nta/results"),
+    ray_trainable=DebugTrainable,
+    algorithm_class=DQN,
+    experiment_class=OffPolicyReinforcementLearningExperiment,
+    dataset_class=EnvDataset,
+    dataset_args=dict(
+        env_init_func=gym_env,
+        env_args=dict(
+            id="Breakout-v4"
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/dqn/experiments/base.py:49]: Won't work automatically. ROMs need to be downloaded. I will add a comment to requirements.txt, if we use this in the future, I can make a Readme with further explanations

File: projects/dqn/experiments/__init__.py [hunk 17]
```diff
@@ -0,0 +1,29 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2022, Numenta, Inc.  Unless you have an agreement
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/dqn/experiments/__init__.py:3]: Not anymore =)

File: projects/dqn/src/data_utils.py [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+from torchvision import transforms
+
+Experience = namedtuple("Experience", (
+    "observation", "action", "reward", "next_observation", "done", "info")
+)
+
+
+class EnvDataset(Dataset):
+    """
+    Based on gym
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply projects/dqn/src/data_utils.py:35]: changed

File: projects/dqn/experiments/base.py (hunk 6 above)
Comment (lucasosouza) [reply projects/dqn/experiments/base.py]: changed to `main` as per discussion

**Review by lscheinkman — APPROVED** [review]
